File: ts/2022/day19.ts
import { main } from 'lib/advent';
import { lines, product, sum } from 'lib/util';

type Resource = 'ore' | 'clay' | 'obsidian' | 'geode';
type Blueprint = {
  id: number;
} & Record<Resource, Vector4>;

type Vector4 = [number, number, number, number];
const vec = {
  add(a: Vector4, b: Vector4): Vector4 {
    return [a[0] + b[0], a[1] + b[1], a[2] + b[2], a[3] + b[3]];
  },
  sub(a: Vector4, b: Vector4): Vector4 {
    return [a[0] - b[0], a[1] - b[1], a[2] - b[2], a[3] - b[3]];
  },
};

function parse(lines: string[]): Blueprint[] {
  return lines.map((line) => {
    const [id, ore, clay, obsidianOre, obsidianClay, geodeOre, geodeObsidian] =
      line.match(/\d+/g).map(Number);
    return {
      id,
      ore: [ore, 0, 0, 0],
      clay: [clay, 0, 0, 0],
      obsidian: [obsidianOre, obsidianClay, 0, 0],
      geode: [geodeOre, 0, geodeObsidian, 0],
    };
  });
}

const RESOURCE = {
  ore: 0,
  clay: 1,
  obsidian: 2,
  geode: 3,
};

function neighbors(
  blueprint: Blueprint,
  storage: Vector4
): [Resource, Vector4][] {
  const rv: [Resource, Vector4][] = [];
  for (const type of ['geode', 'obsidian', 'clay', 'ore'] as Resource[]) {
    const cost = blueprint[type];
    if (vec.sub(storage, cost).every((v) => v >= 0)) rv.push([type, cost]);
  }
  return rv;
}

function maxGeodes(
  blueprint: Blueprint,
  minutes: number,
  maxQueue = 1000
): number {
  type QueueItem = { robots: Vector4; storage: Vector4 };

  function score({ robots, storage }: QueueItem, i: number): number {
    return (
      storage[RESOURCE['geode']] * (minutes - i) * 10000 +
      robots[RESOURCE['obsidian']] * 1000 +
      robots[RESOURCE['clay']] * 100 +
      robots[RESOURCE['ore']]
    );
  }

  const robots: Vector4 = [1, 0, 0, 0];
  const storage: Vector4 = [0, 0, 0, 0];
  let q = [{ robots, storage }];
  for (let i = 0; i < minutes; ++i) {
    if (q.length > maxQueue) {
      q.sort((a, b) => score(b, i) - score(a, i));
      q = q.slice(0, maxQueue);
    }
    const nextq = [];
    while (q.length) {
      const { robots, storage } = q.shift();
      const nextStorage = vec.add(storage, robots);
      const canBuild = neighbors(blueprint, storage);
      for (const [type, cost] of canBuild) {
        const newRobot: Vector4 = [0, 0, 0, 0];
        newRobot[RESOURCE[type]] = 1;
        nextq.push({
          robots: vec.add(robots, newRobot),
          storage: vec.sub(nextStorage, cost),
        });
      }
      nextq.push({ robots, storage: nextStorage }); // build nothing, conserving our resources
    }
    q = nextq;
  }
  return Math.max(...q.map(({ storage }) => storage[RESOURCE['geode']]));
}

// XXX bug somewhere in part 1
main(
  (s) => sum(parse(lines(s)).map((b) => maxGeodes(b, 24) * b.id)),
  (s) =>
    product(
      parse(lines(s))
        .slice(0, 3)
        .map((b) => maxGeodes(b, 32, 2000))
    )
);

File: py/2022/day19.py
from dataclasses import dataclass
from math import prod

from aoc import main
from aoc.parse import all_numbers, line_parser


# Used for inventory of resources and robots.
@dataclass(eq=True, frozen=True, order=True)
class Resource:
    geode: int
    obsidian: int
    clay: int
    ore: int

    def __add__(self, other: "Resource") -> "Resource":
        return Resource(
            self.geode + other.geode,
            self.obsidian + other.obsidian,
            self.clay + other.clay,
            self.ore + other.ore,
        )

    def __sub__(self, other: "Resource") -> "Resource":
        return Resource(
            self.geode - other.geode,
            self.obsidian - other.obsidian,
            self.clay - other.clay,
            self.ore - other.ore,
        )

    def all_lte(self, other: "Resource") -> bool:
        return (
            self.geode <= other.geode
            and self.obsidian <= other.obsidian
            and self.clay <= other.clay
            and self.ore <= other.ore
        )

    def multiply(self, n: int) -> "Resource":
        return Resource(self.geode * n, self.obsidian * n, self.clay * n, self.ore * n)


Option = tuple[Resource, Resource]  # cost, new robots
Blueprint = tuple[int, list[Option]]
QueueItem = tuple[Resource, Resource]  # inventory: resources, robots


@line_parser
def parse(
    line: str,
) -> Blueprint:
    i, a, b, c, d, e, f = all_numbers(line)
    options = [
        (Resource(0, 0, 0, a), Resource(0, 0, 0, 1)),
        (Resource(0, 0, 0, b), Resource(0, 0, 1, 0)),
        (Resource(0, 0, d, c), Resource(0, 1, 0, 0)),
        (Resource(0, f, 0, e), Resource(1, 0, 0, 0)),
        (Resource(0, 0, 0, 0), Resource(0, 0, 0, 0)),  # do nothing
    ]
    return (i, options)


def max_geodes(options: list[Option], time: int, max_q: int):
    def score(i: int, qi: QueueItem) -> Resource:
        return qi[0] + qi[1].multiply(time - i)

    q: list[QueueItem] = [(Resource(0, 0, 0, 0), Resource(0, 0, 0, 1))]
    for i in range(time):
        uniq_q = {score(i, x): x for x in q}.items()
        # hack: drop all but the highest-scored items
        q = [item for _, item in sorted(uniq_q, reverse=True)][:max_q]
        next = []
        for resources, robots in q:
            next_resources = resources + robots
            for cost, new_robots in options:
                if cost.all_lte(resources):
                    next.append((next_resources - cost, robots + new_robots))
        q = next
    return max(resources.geode for (resources, _) in q)


if __name__ == "__main__":
    main(
        lambda s: sum(max_geodes(opts, 24, 50) * i for i, opts in parse(s)),
        lambda s: prod(max_geodes(opts, 32, 500) for _, opts in parse(s)[:3]),
    )
